### ISSUE TITLE
Serialization of maps with non-string key

### ISSUE THREAD

**Issue by patternoia** [danielaparker/jsoncons#196]
Hi,
I would really like to have an `int` keyed map working with jsoncons.
For this I though about converting the keys on the fly to string an back during serialization and deserialization.

However, if I try to create such a trait, I get a conflict with already supported maps. See the code and error log below.  Is there any way out?

Thanks.

```
namespace jsoncons {

    template <class Json, typename T>
    struct json_type_traits<Json,std::map<int, T>> {

        static const bool is_assignable = true;
        static bool is(const Json& val) noexcept {
            if (!val.is_object())
                return false;

            return true;
        }

        static std::map<int, T> as(const Json& val) {
            std::map<std::string, T> map = val.template as<std::map<std::string, T>>();

            std::map<int, T> result(map.size());
            std::transform(std::begin(map), std::end(map),
                   std::begin(result),
                   [](const std::pair<std::string, T>& element) -> std::pair<int, T> {
                      return std::make_pair<int, T>(std::stoi(element.first), element.second);
                    });

            return result;
        }

        static Json to_json(std::map<int, T> val) {
          std::map<std::string, T> result(val.size());
          std::transform(std::begin(val), std::end(val),
                 std::begin(result),
                 [](const std::pair<int, T>& element) -> std::pair<std::string, T> {
                    return std::make_pair<std::string, T>(std::to_string(element.first), element.second);
                  });
          return Json(result);
        }
    };

    namespace detail {
        template<typename T>
        struct is_map_like<std::map<int, T>>: public std::true_type {};
    }
}
```

Error log

```
include/jsoncons/basic_json.hpp:2940:16: error: ambiguous partial specializations of 'json_type_traits<jsoncons::basic_json<char, jsoncons::preserve_order_policy, std::__1::allocator<char> >,
      std::__1::map<int, int, std::__1::less<int>, std::__1::allocator<std::__1::pair<const int, int> > >, void>'
        : var_(json_type_traits<basic_json,T>::to_json(val).var_)
               ^
include/jsoncons/json_container_types.hpp:1511:23: note: in instantiation of function template specialization 'jsoncons::basic_json<char, jsoncons::preserve_order_policy, std::__1::allocator<char>
      >::basic_json<std::__1::map<int, int, std::__1::less<int>, std::__1::allocator<std::__1::pair<const int, int> > > >' requested here
            it->value(Json(std::forward<T>(value)));
                      ^
include/jsoncons/basic_json.hpp:4150:35: note: in instantiation of function template specialization 'jsoncons::json_object<std::__1::basic_string<char>, jsoncons::basic_json<char,
      jsoncons::preserve_order_policy, std::__1::allocator<char> >, void>::insert_or_assign<std::__1::map<int, int, std::__1::less<int>,
      std::__1::allocator<std::__1::pair<const int, int> > > &, std::__1::allocator<char> >' requested here
            return object_value().insert_or_assign(name, std::forward<T>(val));
                                  ^
include/jsoncons/basic_json.hpp:2019:45: note: in instantiation of function template specialization 'jsoncons::basic_json<char, jsoncons::preserve_order_policy, std::__1::allocator<char>
      >::insert_or_assign<std::__1::map<int, int, std::__1::less<int>, std::__1::allocator<std::__1::pair<const int, int> > > &>' requested here
            parent_.evaluate_with_default().insert_or_assign(key_, std::forward<T>(val));
                                            ^
note: in instantiation of function template specialization 'jsoncons::basic_json<char, jsoncons::preserve_order_policy, std::__1::allocator<char>
      >::proxy<jsoncons::basic_json<char, jsoncons::preserve_order_policy, std::__1::allocator<char> > >::operator=<std::__1::map<int, int,
      std::__1::less<int>, std::__1::allocator<std::__1::pair<const int, int> > > &>' requested here
z["a"] = a;
       ^
include/jsoncons/json_type_traits.hpp:707:8: note: partial specialization matches [with Json = jsoncons::basic_json<char, jsoncons::preserve_order_policy, std::__1::allocator<char> >, T =
      std::__1::map<int, int, std::__1::less<int>, std::__1::allocator<std::__1::pair<const int, int> > >]
struct json_type_traits<Json, T, 
       ^
Utils.h:216:12: note: partial specialization matches [with Json = jsoncons::basic_json<char, jsoncons::preserve_order_policy, std::__1::allocator<char> >, T = int]
    struct json_type_traits<Json,std::map<int, T>> {

```

**Comment by danielaparker** [comment]
This is where you need `is_json_type_traits_declared`, subject of other issue. Simplifying a little:

```c++
namespace jsoncons {

template <class Json, typename T>
struct json_type_traits<Json, std::map<int, T>> {

    static const bool is_assignable = true;
    static bool is(const Json& val) noexcept {
        if (!val.is_object())
            return false;

        return true;
    }

    static std::map<int, T> as(const Json& val) {
        std::map<int, T> result;
        for (const auto& item : val.object_range())
        {
            result.emplace(std::stoi(item.key()), item.value().as<T>());
        }

        return result;
    }

    static Json to_json(const std::map<int, T>& val) {
        Json j;
        j.reserve(val.size());
        for (const auto& item : val)
        {
            j.try_emplace(std::to_string(item.first), item.second);
        }
        return j;
    }
};
template <class T>
struct is_json_type_traits_declared<std::map<int,T>> : public std::true_type {};
}

int main()
{
    std::map<int, int> val{ {1,1,},{2,2} };

    jsoncons::json j{ val };

    std::cout << j << "\n\n";

    std::map<int, int> other = j.as<std::map<int, int>>();

    for (const auto& item : other)
    {
        std::cout << item.first << ", " << item.second << "\n";
    }
}
```
Output:
```c++
{"1":1,"2":2}

1, 1
2, 2
```

**Comment by patternoia** [comment]
Awesome!

Is the `is_map_like` from original code correct? Or I do not need to declare that?

**Comment by danielaparker** [comment]
No, the `is_map_like` line should be omitted,  `is_map_like` is an internal helper template in the `detail` namespace that should not be specialized in user code. Also, the line in the original code, 
```c++
static const bool is_assignable = true;
```
does nothing and can be omitted.

**Comment by patternoia** [comment]
Thanks

**Comment by danielaparker** [comment]
Also, you can generalize the specialization to work for any integral type,

```c++
namespace jsoncons {

template <class Json, typename T, typename Integer>
struct json_type_traits<Json, std::map<Integer, T>,
    typename std::enable_if<std::is_integral<Integer>::value
    >::type> {

    static bool is(const Json& val) noexcept {
        if (!val.is_object())
            return false;

        return true;
    }

    static std::map<Integer, T> as(const Json& val) {
        std::map<Integer, T> result;
        for (const auto& item : val.object_range())
        {
            result.emplace(std::stoi(item.key()), item.value().as<T>());
        }

        return result;
    }

    static Json to_json(const std::map<Integer, T>& val) {
        Json j;
        j.reserve(val.size());
        for (const auto& item : val)
        {
            j.try_emplace(std::to_string(item.first), item.second);
        }
        return j;
    }
};
template <class T, class Integer>
struct is_json_type_traits_declared<std::map<Integer, T>> : public std::true_type {};
}

int main()
{
    std::map<unsigned long, int> val{ {1,1,},{2,2} };

    jsoncons::json j{ val };

    std::cout << j << "\n\n";

    std::map<unsigned long, int> other = j.as<std::map<unsigned long, int>>();

    for (const auto& item : other)
    {
        std::cout << item.first << ", " << item.second << "\n";
    }
}
```

**Comment by patternoia** [comment]
But then usage of `std::stoi` is not correct. Maybe the std::stoll would serve the purpose.

**Comment by danielaparker** [comment]
Right! I missed that.

**Comment by patternoia** [comment]
Still it will make two conversions, from json integral to long and to actual type behind Integer

**Comment by patternoia** [comment]
Somehow this new template conflicts with a `map<string,string>`
I think it should be enabled only for integral types, but I can't get it working with typename `std::enable_if<std::is_integral<Integer>::value>::type`

**Comment by danielaparker** [comment]
It shouldn't, as long as you've also templated
```c++
template <class T, class Integer>
struct is_json_type_traits_declared<std::map<Integer, T>> : public std::true_type {};
```
My example works with vs2019. What compiler and OS are you using?

**Comment by danielaparker** [comment]
Reopening, support for integer keyed map is now supported on master for all integer types, with conversion to/from string keys in basic_json.

**Comment by danielaparker** [comment]
Included in v0.143.0
